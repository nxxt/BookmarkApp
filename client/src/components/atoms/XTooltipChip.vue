<template>
  <v-tooltip bottom>
    <template v-slot:activator="{ on }">
      <div
        class="x-tooltip-chip px-2"
        :class="{ ...classes, [color]: true }"
        :style="{ borderRadius: '5px' }"
        v-on="on"
      >
        <slot></slot>
      </div>
    </template>

    <!-- tooltip content -->
    <span>{{ tooltip }}</span>
  </v-tooltip>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

@Component({})
export default class XTooltipChip extends Vue {
  @Prop({ type: [String, Number], default: 'tooltip' })
  tooltip!: string | number

  @Prop({ type: Object, default: () => ({}) })
  classes!: { [key: string]: boolean }

  @Prop({ type: [String, Number], default: 'blue' })
  color!: string | number

  onDuplicateTagName() {
    this.$emit('duplicate-tag-name')
  }
}
</script>
