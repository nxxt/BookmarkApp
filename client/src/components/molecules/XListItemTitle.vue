<template>
  <div class="x-list-item-title d-flex align-center">
    <x-tooltip-chip
      label
      x-small
      :classes="{
        'body-2': true,
        'mr-2': true,
        'white--text': true,
        'font-weight-bold': true
      }"
      :style="{ width: '100px' }"
      :color="$getColorByNumber(number)"
      tooltip="watch#"
    >
      {{ number }}
    </x-tooltip-chip>

    <a
      :class="$style.list_item_title_text"
      class="white--text"
      :title="url"
      :href="url"
      target="_blank"
      @click.stop="$emit('click', _id)"
    >
      {{ text }}
    </a>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import XTooltipChip from '@/components/atoms/XTooltipChip.vue'

@Component({
  components: { XTooltipChip }
})
export default class XListItemTitle extends Vue {
  // objectId
  @Prop({ type: String, default: '_id' })
  _id!: string

  // watchNumber
  @Prop({ type: Number, default: 0 })
  number!: number

  // title
  @Prop({ type: String, default: 'title' })
  text!: string

  // url
  @Prop({ type: String, default: 'url' })
  url!: string
}
</script>

<style module>
.list_item_title_text {
  text-decoration: none;
}
</style>
