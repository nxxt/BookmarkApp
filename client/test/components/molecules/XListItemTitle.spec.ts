import Vuetify from 'vuetify'
import { mount, createLocalVue } from '@vue/test-utils'
import XListItemTitle from '@/components/molecules/XListItemTitle.vue'
import mockGetters from '@/utils/mock_getters'
import XTooltipChip from '@/components/atoms/XTooltipChip.vue'

// plugin 登録
const localVue = createLocalVue()
Object.keys(mockGetters).forEach((key: string) => {
  localVue.prototype[key] = mockGetters[key]
})

describe('XListItemTitle.vue', () => {
  /****************
   * 準備
   ****************/
  // vuetify
  let vuetify: typeof Vuetify
  beforeEach(() => {
    vuetify = new Vuetify()
  })

  // propsで渡す値
  const propsData = {
    _id: 'test_id',
    number: 10,
    text: 'test_text',
    url: 'test_url'
  }

  const factory = (options?: any) => {
    return mount(XListItemTitle, {
      localVue,
      vuetify,
      ...options
    })
  }

  /****************
   * TEST
   ****************/
  it('props', () => {
    const wrapper = factory({ propsData })
    // VueインスタンスのpropsがpropsDataと等しいか
    expect(wrapper.props()).toStrictEqual(propsData)
  })

  it('child component', () => {
    const wrapper = factory({ propsData })
    expect(wrapper.contains(XTooltipChip)).toBe(true)
  })

  describe('tooltip chip', () => {
    it('tooltip chip content', () => {
      const wrapper = factory({ propsData })
      const elm = wrapper.find('.x-tooltip-chip')

      expect(Number(elm.text())).toBe(propsData.number)
    })
  })

  describe('a (linked title)', () => {
    it('attrs and contents', () => {
      const wrapper = factory({ propsData })
      const elm = wrapper.find('a')

      // 初期状態の検証
      expect(elm.attributes().title).toBe(propsData.url)
      expect(elm.attributes().href).toBe(propsData.url)
      expect(elm.text()).toBe(propsData.text)
    })

    it('click event', () => {
      const wrapper = factory({ propsData })
      const event = jest.fn()
      const elm = wrapper.find('a')

      wrapper.vm.$on('click', event) // wrapperでclickイベント検知用

      // 初期状態の検証
      expect(event).toBeDefined() // イベントがundefinedでないこと
      expect(event).toHaveBeenCalledTimes(0) // eventが発生していないこと

      // Simulate a click on the elm
      elm.trigger('click')

      // 発生したイベントの検証
      expect(event).toHaveBeenCalledTimes(1) // イベントが親へ伝搬したこと
      const a = wrapper.emitted('click')
      expect(a).toBeDefined()
      if (a) expect(a[0][0]).toEqual(propsData._id) // イベントのペイロードが正しいこと
    })
  })

  describe('The entire template', () => {
    it('snapshot', () => {
      // コンポーネントラッパーの作成
      const wrapper = factory({ propsData })

      // レンダリング結果が前回のスナップショットと同じか
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
