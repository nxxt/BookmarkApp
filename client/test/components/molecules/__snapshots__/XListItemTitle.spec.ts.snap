// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XListItemTitle.vue The entire template snapshot 1`] = `
"<div class=\\"x-list-item-title d-flex align-center\\"><span class=\\"v-tooltip v-tooltip--bottom\\" label=\\"\\" x-small=\\"\\" style=\\"width: 100px;\\"><!----><div class=\\"x-tooltip-chip px-2 body-2 mr-2 white--text font-weight-bold blue\\" style=\\"border-radius: 5px;\\">
    10
  </div></span> <a title=\\"test_url\\" href=\\"test_url\\" target=\\"_blank\\" class=\\"white--text list_item_title_text\\">
    test_text
  </a></div>"
`;
